File: source150/114_flatten.go
package source150

/**
地址：https://leetcode-cn.com/problems/flatten-binary-tree-to-linked-list/
题目描述：
给定一个二叉树，原地将它展开为链表。

Example:
例如，给定二叉树

    1
   / \
  2   5
 / \   \
3   4   6
将其展开为：

1
 \
  2
   \
    3
     \
      4
       \
        5
         \
          6
 */

 /**
 思路：
 1. 找到i节点左子树的最大数值节点，即找到左子树的最右节点，记为preNode
 2. 将i节点的右子树放到preNode右子树
 3. 将i节点的左子树移动到右子树位置
 4. 递归i节点右子节点
  */

  /**
  用时：0ms
  内存：3MB
   */
func flatten(root *TreeNode)  {
	if root == nil {
		return
	}
	if root.Left == nil {
		flatten(root.Right)
		return
	}
	preNode := root.Left
	for preNode.Right != nil {
		preNode = preNode.Right
	}
	preNode.Right = root.Right
	root.Right = root.Left
	root.Left = nil
	flatten(root.Right)
}

File: source150/114_flatten.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root: return None
        def helper(node: TreeNode) -> TreeNode:
            left_node = node.left
            right_node = node.right
            if not left_node and not right_node: return node
            if not left_node:
                right_last_node = helper(right_node)
                return right_last_node
            if not right_node:
                left_last_node = helper(left_node)
                node.right = left_node
                node.left = None
                return left_last_node
            left_last_node = helper(left_node)
            right_last_node = helper(right_node)
            left_last_node.right = right_node
            left_last_node.left = None
            node.right = left_node
            node.left = None
            return right_last_node

        helper(root)
        return
